src/main/java/zmq/MailboxSafe.java

```diff
@@ -0,0 +1,131 @@
+package zmq;
+
+import zmq.pipe.YPipe;
+import zmq.util.Errno;
+
+import java.io.IOException;
+import java.util.Vector;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.concurrent.locks.Condition;
+
+public class MailboxSafe implements IMailbox
+{
+    //  The pipe to store actual commands.
+    private final YPipe<Command> cpipe;
+
+    //  Synchronize access to the mailbox from receivers and senders
+    private final ReentrantLock sync;
+
+    //  Condition variable to pass signals from writer thread to reader thread.
+    private final Condition condition;
+
+    private final Vector<Signaler> signalers;
+
+    // mailbox name, for better debugging
+    private final String name;
+
+    private final Errno errno;
+
+    public MailboxSafe(Ctx ctx, ReentrantLock sync, String name)
+    {
+        this.errno = ctx.errno();
+        this.cpipe = new YPipe<>(Config.COMMAND_PIPE_GRANULARITY.getValue());
+        this.sync = sync;
+        this.condition = this.sync.newCondition();
+        this.signalers = new Vector<Signaler>(10);
+        this.name = name;
+
+        //  Get the pipe into passive state. That way, if the users starts by
+        //  polling on the associated file descriptor it will get woken up when
+        //  new command is posted.
+        Command cmd = cpipe.read();
+        assert (cmd == null);
+    }
+
+    public void addSignaler(Signaler signaler)
+    {
+        this.signalers.add(signaler);
+    }
+
+    public void removeSignaler(Signaler signaler)
+    {
+        this.signalers.remove(signaler);
+    }
+
+    public void clearSignalers()
+    {
+        this.signalers.clear();
+    }
+
+    @Override
+    public void send(Command cmd)
+    {
+        sync.lock();
+        try {
+            cpipe.write(cmd, false);
+            boolean ok = cpipe.flush();
+
+            if (!ok) {
+                condition.signalAll();
+                signalers.forEach(Signaler::send);
```

Comment:
Java iterators generate garbage unfortunately. Not sure if it matters here

Reply:
it matters, this would be called alot, is this also generate garbage:

```
for (Signaler signaler : signalers) {
    signaler.send();
}
```

or should I use:

```
for (int i = 0; i < signalers.size(); i++) {
    signalers.get(i).send();
}
```

Reply:
the latter